zerver/migrations/0333_realm_import_audit_log_backfill.py

```diff
@@ -0,0 +1,101 @@
+# Generated by Django 3.1.8 on 2021-04-20 10:09
+
+from django.db import migrations
+from django.db.backends.postgresql.schema import DatabaseSchemaEditor
+from django.db.migrations.state import StateApps
+
+
+def backfill_realm_import_log_events(apps: StateApps, schema_editor: DatabaseSchemaEditor) -> None:
+    Message = apps.get_model("zerver", "Message")
+
+    Realm = apps.get_model("zerver", "Realm")
+    Realm.SELF_HOSTED = 1
+    Realm.LIMITED = 2
+
+    RealmAuditLog = apps.get_model("zerver", "RealmAuditLog")
+    RealmAuditLog.REALM_PLAN_TYPE_CHANGED = 204
+    RealmAuditLog.REALM_IMPORTED = 216
+
+    for service in ["slack", "hipchat", "mattermost"]:
+        objects_to_create = []
+        realms_from_service = (
+            Message.objects.filter(subject__iexact=f"imported from {service}")
+            .values_list("sender__realm", flat=True)
+            .distinct()
+        )
```

Comment:
A much faster and potentially cleaner tell would be to look at RealmAuditLog and see if their organization description is "Imported from slack" and/or ; but I'm not sure if RealmAuditLog for that property goes back before the import tools, and/or if the import tools always had the same strings for `realm.description`.

Reply:
> A much faster and potentially cleaner tell would be to look at RealmAuditLog and see if their organization description is "Imported from slack" and/or ; 


You mean we should filter for RealmAuditLogs with `event_type=RealmAuditLog.REALM_PROPERTY_CHANGED` and `extra_data={RealmAuditLog.OLD_VALUE: f"Imported from {service}", "property": "description"}`.

Looks like this audit log was added only in ba92666dbd30e528d6cc23a7bc40c6bc618e252a which merged on on 7 Jul 2020. So I think relying on this is not of much use.

Also, I think we would have to fetch all audit logs of type `RealmAuditLog.REALM_PROPERTY_CHANGED` for a realm, go through them in a loop and check that extra_data. I don't think we can write a query that does that automatically since we are not specifying `RealmAuditLog.NEW_VALUE`. Does @edith007's work on `JSONField` makes queries like this possible?

> if the import tools always had the same strings for realm.description

The commit that properly sets the `realm.description` was added on  Wed Feb 21 14:57:37 2018.
e98dd634151b55a51e7968b8aa8af57487e48cb5. Also it looks like that commit didn't really worked so it looks like 1410a1e46033e32eb8d14eaf979e0f1b16057700 was the commit that actually fixed it on 18 Apr 2018. I think that is also kind of late?

Reply:
yes @hackerkid queries like this will be possible with `JSONField`. With the change, we won't have to fetch all the RealmAuditLogs and loop through each of them to search for a certain extra_data, instead we can give the json or some part of the json or keys in the query itself...